lib/auth/usertoken.go

```diff
@@ -452,6 +465,110 @@ func (s *Server) createRecoveryToken(ctx context.Context, username, tokenType st
 	return newToken, nil
 }
 
+// CreatePrivilegeToken implements AuthService.CreatePrivilegeToken.
+func (s *Server) CreatePrivilegeToken(ctx context.Context, req *proto.CreatePrivilegeTokenRequest) (types.UserToken, error) {
+	username, err := GetClientUsername(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authPref, err := s.GetAuthPreference(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if !authPref.GetAllowLocalAuth() {
+		return nil, trace.AccessDenied("local auth needs to be enabled")
+	}
+
+	secondFactor := authPref.GetSecondFactor()
+	if secondFactor == constants.SecondFactorOff {
+		return nil, trace.AccessDenied("second factor must be enabled")
+	}
+
+	tokenKind := UserTokenTypePrivilege
+
+	// Begin authenticating second factor.
+	switch req.GetExistingMFAResponse().GetResponse().(type) {
+	case nil:
+		// Allows users with no devices to bypass second factor re-auth.
+		devices, err := s.Identity.GetMFADevices(ctx, username, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if len(devices) > 0 {
+			return nil, trace.BadParameter("second factor authentication required")
+		}
+
+		tokenKind = UserTokenTypePrivilegeException
+	default:
+		err := s.WithUserLock(username, func() error {
+			_, err := s.validateMFAAuthResponse(ctx, username, req.GetExistingMFAResponse(), s.Identity)
+			return err
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
```

Comment:
Suggestion: reduced err scope:

```suggestion
		if err := s.WithUserLock(username, func() error {
			_, err := s.validateMFAAuthResponse(ctx, username, req.GetExistingMFAResponse(), s.Identity)
			return err
		}); err != nil {
			return nil, trace.Wrap(err)
		}
```